src/mc_tasks/lipm_stabilizer/StabilizerTask.cpp

```diff
@@ -1128,6 +1151,337 @@ void StabilizerTask::distributeWrench(const sva::ForceVecd & desiredWrench)
+  Eigen::MatrixXd Mcop = Eigen::MatrixXd::Zero(nbEqCstr, nbVariables);
+  Eigen::VectorXd bcop = Eigen::VectorXd::Zero(Mcop.rows());
+
+  // Task to regulate the CoPs under the foot ankle
+  Eigen::MatrixXd McopReg = Eigen::MatrixXd::Zero(nbVariables, nbVariables);
+  Eigen::VectorXd bcopReg = Eigen::VectorXd::Zero(McopReg.rows());
+
+  // Task to regulate the CoPs difference
+  Eigen::MatrixXd McopDiff = Eigen::MatrixXd::Zero(2 * nbReferences, nbVariables);
+  Eigen::VectorXd bcopDiff = Eigen::VectorXd::Zero(McopDiff.rows());
+
+  Eigen::MatrixXd Aineq = Eigen::MatrixXd::Zero(nbIneqCstr, nbVariables);
+  Eigen::VectorXd bineq = Eigen::VectorXd::Zero(Aineq.rows());
+
+  Eigen::Matrix<double, 4, 2> normals; // normals matrix for CoP constraints
+  Eigen::Vector4d offsetLeft = Eigen::Vector4d::Zero();
+  Eigen::Vector4d offsetRight = Eigen::Vector4d::Zero();
+  normals << 1, 0, -1, 0, 0, 1, 0, -1;
+
+  offsetLeft << leftContact.halfLength(), leftContact.halfLength(), leftContact.halfWidth(), leftContact.halfWidth();
+
+  offsetRight << rightContact.halfLength(), rightContact.halfLength(), rightContact.halfWidth(),
+      rightContact.halfWidth();
+
+  Eigen::Vector2d t_lankle_zmp = zmp_ref[0] - lankle.segment(0, 2);
+  const double lankle_rankle = t_lankle_rankle.norm();
+  double d_proj = t_lankle_zmp.dot(t_lankle_rankle.normalized());
+  // The vertical forces are splitted using the ratio obtained between the reference zmp pose and the contact pose;
+  double ratio = clamp(d_proj / lankle_rankle, c_.safetyThresholds.MIN_DS_PRESSURE / fz_tot,
+                       1 - (c_.safetyThresholds.MIN_DS_PRESSURE / fz_tot));
+
+  desiredFzLeft_ = (1 - ratio) * fz_tot;
+  desiredFzRight_ = ratio * fz_tot;
+
+  // We modeled the vertical forces to also follow a 1st order behavior between the reference and the state
+  targetForceLeft.z() = desiredFzLeft_ - measuredFzLeft_delayed * exp(-c_.lambda_CoP_Fz.z() * (delta - t_delay));
+  targetForceLeft.z() /= (1 - exp(-c_.lambda_CoP_Fz.z() * (delta - t_delay)));
+
+  targetForceRight.z() = desiredFzRight_ - measuredFzRight_delayed * exp(-c_.lambda_CoP_Fz.z() * (delta - t_delay));
+  targetForceRight.z() /= (1 - exp(-c_.lambda_CoP_Fz.z() * (delta - t_delay)));
+
+  const double ratio0 = ratio;
+
+  for(Eigen::Index i = 0; i < nbReferences; i++)
+  {
+
+    if(i != 0)
+    {
+
+      t_lankle_zmp = zmp_ref[i] - lankle.segment(0, 2);
+      d_proj = t_lankle_zmp.dot(t_lankle_rankle.normalized());
+      // ratio = 1 : fz on rightfoot, 0 on leftFoot
+      ratio = clamp(d_proj / lankle_rankle, c_.safetyThresholds.MIN_DS_PRESSURE / fz_tot,
+                    1 - (c_.safetyThresholds.MIN_DS_PRESSURE / fz_tot));
+    }
+
+    // Acop convert the CoP reference into the modeled CoP
+    // x be vector cotaining the CoP reference for one foot
+    // Acop * x + cop_0 * e^(-lambda t_i) = cop i in foot frame (left/right)
+    Eigen::MatrixXd Acop = Eigen::MatrixXd::Zero(2, 2 * nbReferences);
```

Comment:
This seems wasteful (allocated for every point in the reference entry and only partly filled each time), either:
- allocate outside the loop (better probably)
- allocate with size `2 * (i + 1)`